src/psyclone/dynamo0p3_builtins.py

```diff
@@ -212,28 +219,38 @@ def _validate(self):
         field_count = 0  # We must have one or more fields as arguments
         spaces = set()   # All field arguments must be on the same space
         for arg in self.arg_descriptors:
-            # Built-ins update fields DoF by DoF and therefore can have
-            # WRITE/READWRITE access
-            if arg.access in [AccessType.WRITE, AccessType.SUM,
-                              AccessType.READWRITE]:
-                write_count += 1
-            if arg.argument_type in LFRicArgDescriptor.VALID_FIELD_NAMES:
-                field_count += 1
-                spaces.add(arg.function_space)
+            # Check valid argument types
             if arg.argument_type not in VALID_BUILTIN_ARG_TYPES:
                 raise ParseError(
                     "In the LFRic API an argument to a built-in kernel "
                     "must be one of {0} but kernel '{1}' has an argument of "
                     "type '{2}'.".format(VALID_BUILTIN_ARG_TYPES, self.name,
                                          arg.argument_type))
+            # Check valid data types (InternalError as the relevant ParseError
+            # is caught in LFRicArgDescriptor)
             if (arg.argument_type in LFRicArgDescriptor.VALID_FIELD_NAMES and
```

Comment:
As the comment suggests, isn't this check done in LFRicArgDescriptor? We shouldn't be repeating it.

Reply:
I followed the logic from `dynamo0p3.py` classes (e.g. scalars, fields) where we do the `ParseError` check in the `LFRicArgDescriptor` and then `InternalError` in the class that takes the argument descriptor to make sure nothing is mangled in between. I am happy to remove the check, but then we should agree on where the PSyclone internal checks are necessary and where they are not.

Reply:
The checks in this `_validate()` routine are there to ensure the metadata conforms to the LFRic API for a built-in kernel (c.f. `DynKern._validate()`). I think we have to assume that the values in the metadata are valid otherwise we'll be forever checking values  whenever we call a new routine. If there was a data type that was valid metadata but *not* valid for a builtin kernel then we would check for that here. At the moment all data types are valid so there's no need to check.

However, when we have code like:

    if some_flag == value1:
        do-something
    else:
        do-something-else

Rather than have the catch-all `else` at the end, we explicitly test that `some_flag` is an allowed value otherwise we raise an exception:

    if some_flag == value1:
        do-something
    elif some_flag == value2:
        do-something-else
    else:
        raise InternalError("oh dear")

Reply:
Ok, I realised I put in the `InternalError` checks for the field and scalar data types in `LFRicScalarArgs` and `LFRic Fields` classes that every kernels uses (including built-ins), so the extra checks here are not necessary. I deleted these checks and the related tests.